### ISSUE TITLE
Prometheus unable to parse Helidon MP Exemplars

### ISSUE THREAD

**Issue by dcoracle** [helidon-io/helidon#6318]
[//]: # "Please remove these comments"
[//]: # "Provide a general summary of the issue in the Title above"

## Environment Details
* Helidon Version: 2.5.6
* Helidon SE or Helidon MP : MP
* JDK version: 17
* OS: Oracle Linux 9
* Docker version (if applicable):

----------

## Problem Description
Exemplars are automatically added to metrics of type 'gauge' for JAX-RS annotated Services.  This results in Prometheus ServiceMonitor failing to parse with the following error:

'metric name base_REST_request_elapsedTime_seconds does not support exemplars'

snip from /metrics EP:

```
# TYPE base_REST_request_elapsedTime_seconds gauge
base_REST_request_elapsedTime_seconds{class="com.oracle.oal.wlp.wbs.HTMLService",method="getContent_java.lang.String_java.lang.String_java.lang.String_int_java.lang.String"} 49.343265107 # {trace_id="b17d175d1bfbcd8f"} 49343265107 1677598231.371000
```

According to the following link, gauges are not supposed to be supported and suffix needs to be either _total or _bucket: https://github.com/prometheus/prometheus/issues/9738 

Reading the specs, it is unclear if this should be the case.  Would like to know a workaround/solution to this

**Comment by tjquinno** [comment]
@dcoracle 

1. Exemplars should be present only if you have added a dependency on `helidon-metrics-trace-exemplar` (so not  automatically). Can you confirm if you added that dependency to your project?
2. When you say the service monitor fails to parse, does it:
   1.  reject the entire `/metrics` output submitted to it,
   2. reject the offending gauge entries, or 
   3. announce (perhaps annoyingly) that it ignores the attempts to attach exemplars to gauges but accept the gauge observations in the `/metrics` output?

**Comment by dcoracle** [comment]
yes, the entire /metrics output is rejected and the Prometheus treats the entire service/pod as 'down'.

**Comment by tjquinno** [comment]
Thanks for the update. 

Considering that exemplars are conveyed as _comments_ (after the `#` in the Prometheus format) that seems very harsh.

**Comment by dcoracle** [comment]
I tried to re-install prometheus with debugging on and don't see it in the logs.  The only way I was able to see the error was through the prometheus UI (i.e. after port-forwarding, it shows up in http://localhost:9090/targets).

Transcribing it:

Endpoint: http://10.42.0.247:8090/metrics
State: DOWN
Error: metric name base_REST_request_elapsedTime_seconds does not support exemplars

For more details, basically when the service is initally up, there are no errors, and the metrics EP works fine.  After the 1st request it generates the exemplar as I noted in my 1st post, and the error appears

**Comment by dcoracle** [comment]
I also ensured prometheus was configured with 

```
--enable-feature=exemplar-storage
```

**Comment by dcoracle** [comment]
I take it back.  I did find something in the logs...

```
$ kubectl logs prometheus-kube-prometheus-stack-prometheus-0 | grep exemplar

ts=2023-03-01T15:52:01.152Z caller=main.go:175 level=info msg="Experimental in-memory exemplar storage enabled"
ts=2023-03-01T15:52:32.126Z caller=scrape.go:1366 level=debug component="scrape manager" scrape_pool=serviceMonitor/default/wlp-browser-service-monitor/0 target=http://10.42.0.247:8090/metrics msg="Append failed" err="metric name base_REST_request_elapsedTime_seconds does not support exemplars"
```

**Comment by tjquinno** [comment]
Thanks for the added information.

Helidon should not by itself be adding exemplars for metrics that the OpenMetrics spec does not say supports exemplars. https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md 

I'm in the process of working on a fix.

That said, I still think it's overly intolerant of the Prometheus service monitor to reject output that contains unexpected exemplars. After all, exemplars are expressed as comments in the output. 

I might suggest you also file an issue against the service monitor. I would absolutely get it if the service monitor logged _warnings_ about such unexpected exemplars and then consumed the data anyway (ignoring, of course, the unexpected exemplar data itself). 

But to outright reject output based on what's in _comments_ does not seem right.

**Comment by roidelapluie** [comment]
Hello,

Prometheus will soon support ingesting exemplars on any metric type.

**Comment by tjquinno** [comment]
@roidelapluie Link, please?

**Comment by roidelapluie** [comment]
This issue was discussed at a dev summit, and a corresponding fix was merged into main recently. You see the Pull request at https://github.com/prometheus/prometheus/pull/11984.

It will be available with Prometheus 2.43 next week.

**Comment by tjquinno** [comment]
@roidelapluie Thanks very much for the info. 

@docoracle What release of Prometheus are you using and would upgrading to 2.43 be an option for you?

**Comment by dcoracle** [comment]
Thanks for this.  Unfortunately, our infra team has 2.35 installed and it is not something that I can control.  (I can control the helidon version though). Prometheus was recently updated from an even older version so it will likely be later in the year before they upgrade again

**Comment by tjquinno** [comment]
@dcoracle I have merged the PR to address this but we have not yet built a release containing it. 

Would you be able to clone the Helidon repository and build the `helidon-2.x` branch locally (it takes only a few minutes) and try your application with that Helidon build to see if it works properly in your environment?

Because you would not be running a released version of Helidon, you should not consider your app using your local build of Helidon suitable for production use. I am looking for a way to verify that these changes allow your app to work with the Prometheus service monitor.

**Comment by dcoracle** [comment]
Confirmed fix:

```$ curl localhost:8090/metrics | grep REST
# TYPE base_REST_request_total counter
# HELP base_REST_request_total The number of invocations and total response time of RESTful resource methods since the start of the server.
base_REST_request_total{class="com.somepackage.HTMLService",method="getContent_java.lang.String_java.lang.String_java.lang.String_int_java.lang.String"} 1 # {trace_id="b2aca53b536b00e0"} 1 1678301677.139000
base_REST_request_elapsedTime_seconds{class="com.somepackage.HTMLService",method="getContent_java.lang.String_java.lang.String_java.lang.String_int_java.lang.String"} 33.536743766
base_REST_request_total{class="com.somepackage.HTMLService",method="test"} 0
base_REST_request_elapsedTime_seconds{class="com.somepackage.HTMLService",method="test"} 0.0
```
I checked the ServiceMonitor and there are no longer any issues.

Thank you so much for the quick turnaround!

Do you have a timeline on when this will be released?  I assume it will be 2.5.7

**Comment by tjquinno** [comment]
Resolved by #6333
